compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp

```diff
@@ -264,32 +177,139 @@ FailureOr<PadPackTiling> PadPackTiling::create(linalg::LinalgOp linalgOp) {
   // Tiling in the K-dimension is done differently TODO(newling)
   // document/reconsider.
 
-  auto maxL1Size = 16 * scaleFactor;
-  uint32_t M1 = findLargestFactor(M, maxL1Size, mPack);
-  uint32_t N1 = findLargestFactor(N, maxL1Size, nPack);
-
-  auto maxL0Size = 64 * scaleFactor;
-  uint32_t M0 = findLargestFactor(M, maxL0Size, M1);
-  uint32_t N0 = findLargestFactor(N, maxL0Size, N1);
+  if (isPackPeel) {
+    // Assume working on a 2x2 AIE array, so the ideal level 1 tile sizes should
+    // be (tileM0/2, tileN0/2). Since packing happens before tiling, and an
+    // extra step is performed to fuse pack ops into the loops, the adjusted
+    // level 1 tile sizes should be (tileM0/2/packedM1, tileN0/2/packedN1).
+    auto maxL1Size = 16 * scaleFactor;
+    uint32_t M1 = findLargestFactor(M / m1Pack, maxL1Size / m1Pack, m1Pack);
+    uint32_t N1 = findLargestFactor(N / n1Pack, maxL1Size / n1Pack, n1Pack);
+
+    auto maxL0Size = 32 * scaleFactor;
+    uint32_t M0 = findLargestFactor(M, maxL0Size, m1Pack * M1);
+    uint32_t N0 = findLargestFactor(N, maxL0Size, n1Pack * N1);
+
+    // In pack-peel pipeline there is only one level of tiling for K dimension,
+    // so set K1 = 0. The packed outer K dimension needs to be 1, so set K0 = 1.
+    uint32_t K1 = 0;
+    uint32_t K0 = 1;
+
+    uint32_t m0Pack = M0;
+    uint32_t n0Pack = N0;
+    uint32_t k0Pack = findLargestFactor(K, maxL1Size);
+
+    return ParameterSetting{M0,     N0,     K0,     M1,     N1,     K1,
+                            m0Pack, n0Pack, k0Pack, m1Pack, n1Pack, k1Pack};
+  } else {
+    // Assume working on a 4x4 AIE array. The tile sizes are chosen empirically
+    // for large GEMM sizes, which are [64*s, 64*s, 256] for the first level and
+    // [16*s, 16*s, 16*s] for the second level, where 's' is the scaling factor
+    // based on the element type's bit width.
+    auto maxL1Size = 16 * scaleFactor;
+    uint32_t M1 = findLargestFactor(M, maxL1Size, m1Pack);
+    uint32_t N1 = findLargestFactor(N, maxL1Size, n1Pack);
+
+    auto maxL0Size = 64 * scaleFactor;
+    uint32_t M0 = findLargestFactor(M, maxL0Size, M1);
+    uint32_t N0 = findLargestFactor(N, maxL0Size, N1);
+
+    uint32_t K1 = findLargestFactor(K / k1Pack, 2 * scaleFactor);
+    uint32_t K0 = findLargestFactor(K, 256, k1Pack * K1);
+
+    // In pad-pack pipeline, there is only one level of packing, set pack
+    // parameters for level 0 as 0.
+    uint32_t m0Pack = 0;
+    uint32_t n0Pack = 0;
+    uint32_t k0Pack = 0;
+
+    // Tiling at level 0 (shared memory) should be no smaller than tiling at
```

Comment:
I guess this assertion is not true on the pack-peel pipeline now, can you explain why?